Fixes issue #10184.

davidrjenni opened the pull request:
Fixes #10184.
Merge user specified devices correctly with default devices.
Otherwise the user specified devices end up without permissions.

Signed-off-by: David R. Jenni david.r.jenni@gmail.com

thaJeztah commented:
ping @timthelion @cpuguy83 @calavera

calavera commented:
code LGTM. @davidrjenni do you think it would be possible to provide a test case to avoid future regressions?

calavera commented:
Can we do something like this to merge devices to avoid that On^2 loop?

```go
paths := map[string]*config.Device{}
for _, d := range userDevices {
  paths[d.Path] = d
}

var devs []*config.Device
for _, d := range defaultDevices {
  if _, defined := paths[d.Path]; !defined {
    devs = append(devs, d)
  }
}

return append(devs, userDevices)
```

davidrjenni commented:
@calavera, sure we can avoid the O(n^2) loop.

Would you leave the first if statement?
```
if len(userSpecifiedDevices) == 0 {
	return defaultDevices
}
```
 I think it covers most of the cases in which we would loop unnecessarily.

Do you think the comments I made in the `mergeDevices` function bring any benefit?
Reading them now, I'm in favor of removing them. 

Regarding the tests: in which file should I put the tests?
I guess the test file should only build if `GOOS=linux` (because of the `// +build linux` in `container_linux.go`).
So should I put the tests in a new file, say `container_linux_unit_test.go`, with `// +build linux` on top?

calavera commented:
> Would you leave the first if statement?

yes, that looks good to me.

> So should I put the tests in a new file, say container_linux_unit_test.go, with // +build linux on top?

yes, that's a good option. You can also write an integration-cli test if you feel more conformable with them.

timthelion commented on daemon/container_linux.go:
```diff
@@ -306,6 +307,24 @@ func populateCommand(c *Container, env []string) error {
+	var devs []*configs.Device
+	for _, defaultDevice := range defaultDevices {
+		for _, userSpecified := range userSpecifiedDevices {
+			if defaultDevice.Path != userSpecified.Path {
```
I'm confused, but maybe it's because I don't know go. Does this add a copy of each default device for every user specified device which doesn't have the same path as the default device? If so, that would be wrong. But then again, I really don't know golang so...

davidrjenni replied:
It adds all `defaultDevices` except those which are in `userSpecifiedDevices`. In the last line it adds the `userSpecifiedDevices`.

It merges `defaultDevices` and `userSpecifiedDevices`, but prefers a device in `userSpecifiedDevices` over a device in `defaultDevices`.
Say, there is `{"/dev/fuse", "/dev/null"}`in `defaultDevices` and `{"/dev/fuse"}`in `userSpecifiedDevices`, you get `{"/dev/null", "/dev/fuse"}` where `"/dev/fuse"` is taken from `userSpecifiedDevices`.

Does this make sense?

timthelion replied:
I understand what you are trying to do, but I don't get how your code does this. My python adled brain is reading this as:

````
for defaultDevice in defaultDevices:
  for userSpecified in userSpecifiedDevices:
    if defaultDevice.Path != userSpecified.Path:
      devs = append(devs, defaultDevice)
````

Which would be wrong.

davidrjenni replied:
Yes, this would be wrong.
But I think if we take the last line of `mergeDevices` into account:
```return append(devs, userSpecifiedDevices...)```
it does what I tried to describe.

calavera commented:
The new code is correct, although it needs a rebase with master. Thanks for adding the test.

LGTM.

timthelion commented on daemon/container_linux.go:
```diff
@@ -306,6 +307,25 @@ func populateCommand(c *Container, env []string) error {
+			devs = append(devs, d)
+		}
+	}
+	return append(devs, userDevices...)
```
I think that the current code is clear and correct. Did it occure to you to do it the other way around and write first the default devices into the map and then overwrite the map using the userDevices?

````
paths := map[string]*configs.Device{}
// First add the default devices to the path map.
for _, d := range defaultDevices {
 paths[d.Path] = d
}
//Then over-ride those defaults when necessary.
for _, d := range userDevices {
 paths[d.Path] = d
}
var devs []*configs.Device
for _, d := range paths {
  devs = append(devs,d)
}
return devs
````

I'm actually not sure if this works in golang, it's the pythonic way of doing this though.

calavera replied:
it should work too, I don't think a third loop provides any additional value though. I only wanted to get rid of that O(n^2) :smile:

calavera commented:
LGTM

calavera commented:
@jfrazelle I think it would be good to have this in the release, and it's just a LGTM away from merging :smile:

cpuguy83 commented on integration-cli/docker_cli_daemon_test.go:
```diff
@@ -1165,3 +1165,18 @@ func (s *DockerDaemonSuite) TestRunContainerWithBridgeNone(c *check.C) {
 }
+
+// Issue #10184.
+func (s *DockerDaemonSuite) TestDevicePermissions(c *check.C) {
```
This shouldn't be a daemon test.

cpuguy83 commented:
Just need to move the test from the `DockerDaemonSuite` the normal `DockerSuite`, and put it in `integration-cli/docker_cli_run_test.go`

cpuguy83 commented:
Otherwise LGTM

davidrjenni committed:
Fix issue #10184.

Merge user specified devices correctly with default devices.
Otherwise the user specified devices end up without permissions.

Signed-off-by: David R. Jenni <david.r.jenni@gmail.com>

cpuguy83 commented on integration-cli/docker_cli_run_test.go:
```diff
@@ -3191,3 +3191,18 @@ func (s *DockerSuite) TestRunPublishPort(c *check.C) {
+// Issue #10184.
+func (s *DockerDaemonSuite) TestDevicePermissions(c *check.C) {
+	const permissions = "crw-rw-rw-"
+	if err := s.d.StartWithBusybox(); err != nil {
```
drop this

cpuguy83 commented on integration-cli/docker_cli_run_test.go:
```diff
@@ -3191,3 +3191,18 @@ func (s *DockerSuite) TestRunPublishPort(c *check.C) {
+	if err := s.d.StartWithBusybox(); err != nil {
+		c.Fatalf("could not start daemon with busybox: %v", err)
+	}
+	out, err := s.d.Cmd("run", "--device", "/dev/fuse:/dev/fuse:mrw", "busybox:latest", "ls", "-l", "/dev/fuse")
```
change this to `out, err := dockerCmd(c, "run", "--device", "/dev/fuse:/dev/fuse:mrw", "busybox", "ls", "-l", "/dev/fuse")`

cpuguy83 commented:
LGTM with the new test, thanks!

thaJeztah commented:
ping @calavera @jfrazelle for one more look and cherry-pick for 1.7

calavera commented:
Merging! thank you!

jfrazelle commented:
cherry-picked

calavera commented:
:metal:

davidrjenni commented:
Thank you all for reviewing.
